crypto/keys/keybase_base.go

```diff
@@ -34,6 +48,39 @@ type (
 	}
 )
 
+type optionFunc func(*kbOptions)
+
+func (f optionFunc) apply(o *kbOptions) {
+	f(o)
+}
+
+// WithKeygenFunc applies an overridden key generation function to generate the private key
+func WithKeygenFunc(f PrivKeyGenFunc) KeybaseOption {
+	return optionFunc(func(o *kbOptions) {
+		o.keygenFunc = f
+	})
+}
+
+// newBaseKeybase generates the base keybase with defaulting to tendermint SECP key type
+func newBaseKeybase(opts ...KeybaseOption) baseKeybase {
+	// Default options for keybase
+	options := kbOptions{
+		keygenFunc: baseSecpPrivKeyGen,
+	}
+
+	for _, o := range opts {
+		o.apply(&options)
+	}
+
+	return baseKeybase{
+		options: options,
+	}
+}
```

Comment:
`/baseapp/baseapp.go` provides a good example of this pattern as well:

```go
func NewBaseApp(
	name string, logger log.Logger, db dbm.DB, txDecoder sdk.TxDecoder, options ...func(*BaseApp),
) *BaseApp {
	...
	for _, option := range options {
		option(app)
	}
	...
	return app
}
```

i.e `KeybaseOption` should be a function as follows:

```go
func newBaseKeybase(optionsFns ...KeybaseOption) baseKeybase {
	// Default options for keybase
	 options := kbOptions{keygenFunc: baseSecpPrivKeyGen}

	for _, optionFn := range optionsFns {
		optionFn(&options)
	}
	return baseKeybase{options: options}
}
```

Reply:
ah I hadn't noticed that. Yes I do like this better in this case

Reply:
One last point about this, would it be preferred to make the `kbOptions` struct public and reference the options as `func(*KbOptions)` or do you like the KeybaseOption type definition for readability?

Reply:
the type is ok